File: array/0229_majority_element_ii/0229_majority_element_ii.go
package main 

import (
	"fmt"
)

func majorityElement(nums []int) []int {
	count1, count2, candidate1, candidate2 := 0, 0, 0, 1
	for _, num := range nums {
		if num == candidate1 {
			count1++
		} else if num == candidate2 {
			count2++
		} else if count1 <= 0 {
			candidate1, count1 = num, 1
		} else if count2 <= 0 {
			candidate2, count2 = num ,1
		} else {
			count1--
			count2--
		}
	}

	count1, count2 = 0, 0
	for _, num := range nums {
		if num == candidate1 {
			count1++
		} else if num == candidate2{
			count2++
		}
	}

	length := len(nums)
	if count1 > length/3 && count2 > length/3 {
		return []int{candidate1, candidate2}
	}

	if count1 > length/3 {
		return []int{candidate1}

	}

	if count2 > length / 3 {
		return []int{candidate2}
	}

	return []int{}
}

func main() {
	nums := []int{1,1,1,3,3,2,2,2}
	res := majorityElement(nums)
	fmt.Println(res)
}

File: array/0229_majority_element_ii/0229_majority_element_ii.py
class Solution(object):
    def majorityElement(self, nums):
        if not nums:
            return []

        num1 = -1
        num2 = -1
        count1 = 0
        count2 = 0

        for n in nums:
            if n == num1:
                count1 += 1
            elif n == num2:
                count2 += 1
            elif count1 == 0:
                num1 = n
                count1 = 1
            elif count2 == 0:
                num2 = n
                count2 = 1
            else:
                count1 -= 1
                count2 -= 1
        
        return [n for n in [num1, num2] if nums.count(n) > len(nums) // 3]

nums = [3,2,3]
res = Solution().majorityElement(nums)
print(res)